vote/src/vote_account.rs

```diff
@@ -185,15 +215,28 @@ impl VoteAccounts {
         if stake == 0u64 {
             return;
         }
+
         let Some(staked_nodes) = self.staked_nodes.get_mut() else {
             return;
         };
-        if let Some(node_pubkey) = vote_account.node_pubkey().copied() {
+
+        VoteAccounts::do_add_node_stake(staked_nodes, stake, vote_account.node_pubkey());
+    }
+
+    fn do_add_node_stake(
+        staked_nodes: &mut Arc<HashMap<Pubkey, u64>>,
+        stake: u64,
+        node_pubkey: Option<&Pubkey>,
+    ) {
+        node_pubkey.map(|node_pubkey| {
             Arc::make_mut(staked_nodes)
-                .entry(node_pubkey)
+                // Note we take an Option<&Pubkey> instead of taking by value, which is usually a
+                // bit of an anti pattern, but none of the callers of this function have an owned
+                // Option<Pubkey> so effectively this avoids a copy when node_pubkey is None.
```

Comment:
In practice it's impossible for `node_pubkey` to be None since we always check if the vote account is initialized before storing it in the cache.

Reply:
If you're referring to `VoteStateVersions::is_correct_size_and_initialized(account.data())` called in `StakesCache::check_and_store()`I don't think that's exhaustive? It checks some sizes but deserialization could still fail I think? And because VoteAccount deserializes lazily (which I don't like, but it's what it is), by the time we call vote_account.node_pubkey() deserialization could fail and we get None?

Reply:
Yeah I was referring to the initialization check inside `is_correct_size_and_initialized`. If the vote state being saved is initialized it surely must be deserializable.

Reply:
This is the check

```
    pub fn is_correct_size_and_initialized(data: &[u8]) -> bool {
        const VERSION_OFFSET: usize = 4;
        const DEFAULT_PRIOR_VOTERS_END: usize = VERSION_OFFSET + DEFAULT_PRIOR_VOTERS_OFFSET;
        data.len() == VoteState::size_of()
            && data[VERSION_OFFSET..DEFAULT_PRIOR_VOTERS_END] != [0; DEFAULT_PRIOR_VOTERS_OFFSET]
    }
```

do you agree that this check can return true, but then deserialization can fail because some fields are malformed? for example if the serialized length of VoteState::votes is set to u32::MAX. Are you saying that it's impossible for a malicious actor to craft such a vote and for it to reach this point in the code? I'm not sure about that, I'm fairly sure when I started working on this I saw deserialization errors, but I'll double check!

Reply:
Yeah I'm quite certain, a bug like that would be a security concern

Reply:
Ok cool. I've reverted do_add_node_stake to take the key by value then, since that is more idiomatic.